### ISSUE TITLE
scipy.optimize: threshold on objective function value for early stop.

### ISSUE THREAD

**Issue by Alessi0X** [scipy/scipy#12103]
**Is your feature request related to a problem? Please describe.**
Not a problem per se. It's just a feature request.
The idea is to halt `scipy.optimize` (e.g., `differential_evolution`) when a user-defined threshold on the objective function is reached. Apparently, a workaround can be employed on `basinhopping` (see [here](https://stackoverflow.com/questions/48191589/scipy-basinhopping-with-callback-stop-the-algorithm-when-a-condition-is-met)), but only because `basinhopping` has a different syntax for the `callable` argument.

**Describe the solution you'd like**
Equip the `callable` function with suitable arguments (e.g., the best objective function value obtained so far). In this manner, the user can compare it against a user-defined threshold and trigger a true/false return for halting the evolution beforehand. This can be accomplished on time limits (see [here](https://stackoverflow.com/questions/60854268/stop-scipy-differential-evolution-after-a-time-threshold/60863988#60863988) for halting `differential_evolution` after a user-defined time limit), but not on the objective function value.
Alternatively, the definition of a `threshold` input argument to the optimisation routine (e.g., `differential_evolution`) that automatically takes care of checking (likely at the end of each generation if `update` is `deferred` or after each individual if `update` is `immediate`) whether the condition `best_fitness<=threshold` is verified.

**Describe alternatives you've considered**
None.

**Additional context (e.g. screenshots)**
None.

**Comment by andyfaff** [comment]
TL:DR - it's already possible with existing callback mechanism


An idea that has been expressed in the past is to change the user callback function to accept an intermediate `OptimizeResult` (or `OptimizeState`) instead of the callback function accepting `x`. 
i.e.
```
state = scipy.optimize.OptimizeState(x=x, fun=fun)
callback(state)
# instead of
callback(x)
```

The `OptimizeState` would have the parameter vector/function value, and any other relevant information that might be useful. The difficulty is transitioning from the current type of callback to the new type. This would aid the user `callback` in determining the state of optimization (e.g.  stopping iteration on a user defined function threshold).

However, even if the user callback is not supplied with `fun(x)` the existing way of using callback functions can be used to do the same thing, it's just more of a hassle. e.g. the function value can be recalculated, or obtained from a cached value on the objective function:

```
class Objective():
    def __init__(fun):
        self.fun = fun
        self.best_x = None
        self.best_func = np.inf

    def __call__(self, x,  *args, **kwds):
        fval = self.fun(x, *args, **kwds)
        if fval < self.best_func:
            self.best_func = fval
            self.best_x = x
        return fval
```

The user callback then uses function value to decide when to stop.
